docs/rfcs/0561-list-metadata-reuse.md

```diff
@@ -0,0 +1,131 @@
+- Proposal Name: `list_metadata_reuse`
+- Start Date: 2022-08-23
+- RFC PR: [datafuselabs/opendal#561](https://github.com/datafuselabs/opendal/pull/561)
+- Tracking Issue: None
+
+# Summary
+
+Extend `DirEntry` with more `metadata` fields.
+
+# Motivation
+
+Users may expect to browse metadata of some directories' child files and directories. `list()` seems to be an ideal way to complete this job. 
+
+Thus, they start iterating on it, but soon they realized the `DirEntry`, which could only offer the name (or path, more precisely) and access mode of the object.
+
+So they have to call `metadata()` for each name they extracted from the iterator.
+
+The final example looks like:
+
+```rust
+use anyhow::Result;
+use futures_lite::StreamExt;
+use opendal::ops::OpList;
+use opendal::ops::OpStat;
+use opendal::services::gcs::Builder;
+use opendal::Accessor;
+
+#[tokio::main]
+async fn main() -> Result<()> {
+    let list_op = OpList::new("/path/to/dir")?;
+    let accessor = Builder::default()
+        .bucket("example")
+        .root("/example/root/")
+        .credential("example-credential")
+        .build()?;
+
+    // here is a network request
+    let mut dir_stream = accessor.list(&list_op).await?;
+
+    while let Some(Ok(file)) = dir_stream.next().await {
+        let path = file.path();
+        let stat_op = OpStat::new(path)?;
+
+        // here is another network request
+        let size = accessor.stat(&stat_op).await?.content_length();
+
+        println!("size of file {} is {}B", path, size);
+    }
+    Ok(())
+}
+```
+
+But...wait! many storage-services returns object metadata when listing, like HDFS, AWS and GCS. The rust standard library
+returns metadata when listing local file systems, too.
+
+In the previous versions of OpenDAL we just simply ignored them. This wastes users' time on requesting on metadata.
+
+# Guide-level explanation
+
+The loop in main will be changed to the following code with this RFC:
+```rust
+while let Some(Ok(file)) = dir_stream.next().await {
+    let size = file.len().await;
```

Comment:
1. Keep the same API style with `ObjectMetadata`. If you have different ideas about this, please raise a comment instead of repeating the same code again.
2. We don't want `await` again: users could fetch the meta without extra API call if it already got during `list`.